# ntStat

# Subcommands

## `filter`

Filters *k*-mers based on their counts and stores the output in a Bloom filter (or counting Bloom filter if `--counts` is set, which will make the output 8x larger). *k*-mers with counts between `-cmin` and `-cmax` (inclusive) will be included in the output.

### Usage

```
filter [--help] [--version] [-k VAR] [-s VAR] -f VAR [-e VAR] [-b VAR] [-cmin VAR] [-cmax VAR] [--counts] [--long] [-t VAR] -o VAR reads

Positional arguments:
  reads          path to sequencing data file(s) [nargs: 0 or more] [required]

Optional arguments:
  -h, --help     shows help message and exits
  -v, --version  prints version information and exits
  -k             k-mer length
  -s             path to spaced seeds file (one per line, if -k not specified)
  -f             path to k-mer spectrum file (from ntCard) [required]
  -e             target output false positive rate [nargs=0..1] [default: 0.0001]
  -b             output BF/CBF size (bytes)
  -cmin          minimum count threshold (>=1) [nargs=0..1] [default: 1]
  -cmax          maximum count threshold (<=255) [nargs=0..1] [default: 255]
  --counts       output counts (requires ~8x RAM for CBF)
  --long         optimize for long read data
  -t             number of threads [nargs=0..1] [default: 1]
  -o             path to store output file [required]
```

- Set `-s` to count spaced seeds instead of *k*-mers. `-k` will be ignored in this case. The file pointed to by `-s` should be a plain text file containing spaced seed patterns, one per line, with `1` for care and `0` for don't care positions.
- [TODO] Use `-w` to find minimizer counts. For each consecutive `-w` *k*-mers/spaced seeds, the element with the smallest hash value represents the window's minimizer.
- To automatically set `-cmin` to filter erroneous (weak) *k*-mers and find 'solid' ones, use `--solid`. If `--solid` is set, `-cmin` will be ignored.
- [TODO] Same as `--solid`, use `--repeat` to detect a repeat threshold for `-cmax`.
- `-f` should point to the histogram file generated by [ntCard](https://github.com/bcgsc/ntCard). Remember to use the same `-k` value for ntCard and ntStat. For spaced seeds, assign the seed length to ntCard's `-k`.
- Use `-b` to manually set the output data structure's size. Alternatively, set `-e` to the desired false positive rate (0.0001 by default) and ntStat will automatically calculate a suitable value for `-b`.

## `query`

Converts information from a (counting) Bloom filter to TSV format.

### Usage

- The input data can be any FASTA, FASTQ, BAM, SAM, etc. file (refer to btllib's SeqReader documentation). To read from stdin, use `-`.
- If the data structure contains spaced seed information, pass the path to the spaced seeds text file to `-s`.

```
query [-s VAR] -b VAR -o VAR [--distinct] query

Positional arguments:
  query       path to query data [nargs: 0 or more] [required]

Optional arguments:
  -s          path to spaced seeds file (one per line)
  -b          path to BF/CBF file [required]
  -o          path to output TSV file [required]
  --distinct
```

## `rarity`

Calculate TF-IDF information from sequencing data. Two counting Bloom filters will be stored. One for *k*-mer/seed counts (TF) and other for depths (DF). The number of reads ($N$) will also be printed to stdout. To compute the TF-IDF of an element with $h$ as its hash array, use this formula:

$ \text{TF-IDF}(h) = TF(h) * log \dfrac{D}{DF(h)} $

### Usage

```
rarity [--help] [--version] [-k VAR] [-s VAR] -f VAR [-e VAR] [-b VAR] [--long] [-t VAR] -o VAR reads

Positional arguments:
  reads          path to sequencing data file(s) [nargs: 0 or more] [required]

Optional arguments:
  -h, --help     shows help message and exits
  -v, --version  prints version information and exits
  -k             k-mer length
  -s             path to spaced seeds file (one per line, if -k not specified)
  -f             path to k-mer spectrum file (from ntCard) [required]
  -e             target output false positive rate [nargs=0..1] [default: 0.0001]
  -b             output CBF size (bytes)
  --long         optimize for long read data
  -t             number of threads [nargs=0..1] [default: 1]
  -o             prefix for storing output counting bloom filters [required]
```

## `hist`

### Usage

```
hist [--help] [--version] histogram

Positional arguments:
  histogram      path to k-mer spectrum file (from ntCard) [required]

Optional arguments:
  -h, --help     shows help message and exits
  -v, --version  prints version information and exits
```